File: src/main/java/io/github/tanguygab/tabadditions/shared/features/chat/emojis/EmojiCategory.java
package io.github.tanguygab.tabadditions.shared.features.chat.emojis;

import me.neznamy.tab.shared.platform.TabPlayer;
import java.util.Map;

public record EmojiCategory(String name, Map<String, String> emojis, String output) {

    public boolean canUse(TabPlayer p) {
        return p.hasPermission("tabadditions.chat.emoji.category." + name);
    }

    public boolean canUse(TabPlayer p, String emoji) {
        return canUse(p) || p.hasPermission("tabadditions.chat.emoji." + emoji);
    }

    public int ownedEmojis(TabPlayer p) {
        if (canUse(p)) return emojis.size();
        int i = 0;
        for (String emoji : emojis.keySet())
            if (canUse(p, emoji))
                i++;
        return i;
    }
}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/chat/emojis/EmojiCategory.kt
package io.github.tanguygab.tabadditions.shared.features.chat.emojis

import me.neznamy.tab.shared.platform.TabPlayer

data class EmojiCategory(val name: String, val emojis: MutableMap<String, String>, val output: String) {
    fun canUse(player: TabPlayer) = player.hasPermission("tabadditions.chat.emoji.category.$name")
    fun canUse(player: TabPlayer, emoji: String) = canUse(player) || player.hasPermission("tabadditions.chat.emoji.$emoji")

    fun ownedEmojis(p: TabPlayer) = if (canUse(p)) emojis.size
    else emojis.keys.count { canUse(p, it) }
}